### PR TITLE
SelectMulti

### PR DESCRIPTION
**Pull request by mdodgelooker**
### :sparkles: Changes

- Add `SelectMulti`, which implements `ComboboxMulti`
- `SelectMulti` has the following features not available in `Select`:
  - Multiple `values`
  - `closeOnSelect` (defaults to false)
  - `removeOnBackspace` (added to support the core product's implementation of `react-select`, where it's set to `false` in one usage)
  - `showCreate` & `formatCreateLabel`
- Add `noOptionsLabel` to both `Select` and `SelectMulti`

### :white_check_mark: Requirements

- [x] Includes test coverage for all changes
- [x] Build and tests are passing

### :camera: Screenshots
select-multi

### CODE REVIEW

File: packages/components/package.json [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "@looker/design-tokens": "^0.7.24",
     "@looker/icons": "^0.7.25",
     "@popperjs/core": "^2.1.1",
+    "btoa": "^1.2.1",
```
Comment (mdodgelooker) [inline packages/components/package.json:31]: Fixes the gatsby prod issue with the `Select` caret down icon not showing – it uses `window.btoa`, which isn't available in server side rendering.

File: packages/components/src/Form/Inputs/Combobox/Combobox.tsx [hunk 2]
```diff
@@ -132,26 +133,21 @@ export const ComboboxInternal = forwardRef(
...
 
-    const isVisible = useComboboxToggle(state, onOpen, onClose, option)
+    const isVisible = useComboboxToggle(state, option, onOpen, onClose)
 
     const { ref, ...commonRefs } = useComboboxRefs(forwardedRef)
 
-    // track and share dom information in context.
-    const [listScrollPosition, setListScrollPosition] = useState(0)
-    const [listClientRect, setListClientRect] = useState<DOMRect>()
+    const scrollState = useScrollState()
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/Combobox.tsx:140]: Moving this logic to a custom hook to be shared by `ComboboxMulti`.

File: packages/components/src/Form/Inputs/Combobox/ComboboxContext.ts [hunk 3]
```diff
@@ -62,7 +62,8 @@ export interface ComboboxContextProps<
   transition?: TTransition
   id?: string
   autoCompletePropRef?: MutableRefObject<boolean>
-  persistSelectionRef?: MutableRefObject<boolean>
+  persistSelectionPropRef?: MutableRefObject<boolean>
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxContext.ts:65]: Name change for consistency.

File: packages/components/src/Form/Inputs/Combobox/ComboboxInput.tsx [hunk 4]
```diff
@@ -101,7 +102,7 @@ export const ComboboxInputInternal = forwardRef(
     const isControlled = controlledValue !== undefined
 
     function handleClear() {
-      contextOnChange && contextOnChange()
+      contextOnChange && contextOnChange(undefined)
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxInput.tsx:105]: Changed to make the linter happy.

File: packages/components/src/Form/Inputs/Combobox/ComboboxMulti.tsx [hunk 5]
```diff
@@ -115,12 +116,16 @@ export const ComboboxMultiInternal = forwardRef(
 
     const id = useID(propsID)
 
-    const isVisible = useComboboxToggle(state, onOpen, onClose, options)
+    const isVisible = useComboboxToggle(state, options, onOpen, onClose)
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxMulti.tsx:119]: Making the linter happy.

File: packages/components/src/Form/Inputs/Combobox/ComboboxMulti.tsx [hunk 6]
```diff
@@ -115,12 +116,16 @@ export const ComboboxMultiInternal = forwardRef(
...
 
     const { ref, ...commonRefs } = useComboboxRefs(forwardedRef)
+
+    const scrollState = useScrollState()
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxMulti.tsx:123]: Adding scroll logic added in the `InputTimeSelect` improvements.

File: packages/components/src/Form/Inputs/Combobox/ComboboxMultiOption.tsx [hunk 7]
```diff
@@ -67,21 +76,32 @@ const ComboboxMultiOptionInternal = forwardRef(
       value
     )
 
+    const scrollRef = useOptionScroll(
+      ComboboxMultiContext,
+      value,
+      label,
+      scrollIntoView,
+      isActive
+    )
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxMultiOption.tsx:85]: Adding scroll logic added in the `InputTimeSelect` improvements.

File: packages/components/src/Form/Inputs/Combobox/ComboboxMultiOption.tsx [hunk 8]
```diff
@@ -67,21 +76,32 @@ const ComboboxMultiOptionInternal = forwardRef(
...
-              <CheckMark />
-            </FauxCheckbox>
-          </CheckboxContainer>
+          {!hideCheckMark && (
+            <CheckboxContainer checked={isSelected}>
+              <FauxCheckbox>
+                <CheckMark />
+              </FauxCheckbox>
+            </CheckboxContainer>
+          )}
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxMultiOption.tsx:102]: `hideCheckMark` is used in the "Create" option, where a check mark doesn't look right.

File: packages/components/src/Form/Inputs/Combobox/ComboboxOption.tsx [hunk 9]
```diff
@@ -83,19 +84,6 @@ export interface HighlightTextProps {
...
-  listElement: HTMLLIElement,
-  containerScrollPosition: number,
-  containerHeight = 0
-) => {
-  const { offsetTop } = listElement
-  const isAbove = offsetTop < containerScrollPosition
-  const isBelow = offsetTop > containerScrollPosition + containerHeight
-  return (isAbove && 'above') || (isBelow && 'below') || 'visible'
-}
-
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/ComboboxOption.tsx:98]: Moving all the scroll logic to a custom hook to share with `ComboboxOptionMulti`.

File: packages/components/src/Form/Inputs/Combobox/utils/state.ts [hunk 10]
```diff
@@ -150,44 +150,44 @@ export const stateChart: StateChart = {
...
+        [ComboboxActionType.CHANGE_VALUES]: ComboboxState.SUGGESTING,
         [ComboboxActionType.FOCUS]: ComboboxState.SUGGESTING,
         [ComboboxActionType.NAVIGATE]: ComboboxState.NAVIGATING,
         [ComboboxActionType.CLEAR]: ComboboxState.IDLE,
         [ComboboxActionType.ESCAPE]: ComboboxState.IDLE,
         [ComboboxActionType.BLUR]: ComboboxState.IDLE,
-        [ComboboxActionType.SELECT_WITH_CLICK]: ComboboxState.IDLE,
-        [ComboboxActionType.SELECT_SILENT]: ComboboxState.IDLE,
+        [ComboboxActionType.SELECT_WITH_CLICK]: ComboboxState.SUGGESTING,
+        [ComboboxActionType.SELECT_SILENT]: ComboboxState.SUGGESTING,
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/state.ts:160]: No longer closing the list on select (multi only).

File: packages/components/src/Form/Inputs/Combobox/utils/state.ts [hunk 11]
```diff
@@ -207,7 +207,8 @@ const findNavigationValue = (
     return singularState.option
       ? singularState.option
       : multiState.options
-      ? multiState.options[0]
+      ? multiState.navigationOption ||
+        multiState.options[multiState.options.length - 1]
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/state.ts:211]: Improve the `persistSelection` behavior for multi.

File: packages/components/src/Form/Inputs/Combobox/utils/useComboboxRefs.ts [hunk 12]
```diff
@@ -50,12 +50,14 @@ export function useComboboxRefs(forwardedRef: Ref<HTMLDivElement>) {
...
-  const persistSelectionRef = useRef(false)
+  const persistSelectionPropRef = useRef(false)
+  const closeOnSelectPropRef = useRef(true)
 
   return {
     autoCompletePropRef,
+    closeOnSelectPropRef,
     optionsRef,
-    persistSelectionRef,
+    persistSelectionPropRef,
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/useComboboxRefs.ts:60]: Name change for consistency.

File: packages/components/src/Form/Inputs/Combobox/utils/useComboboxToggle.ts [hunk 13]
```diff
@@ -48,9 +48,9 @@ export function useComboboxToggle<
   TOption extends ComboboxOptionType = MaybeComboboxOptionObject
 >(
   state: ComboboxState,
+  option: TOption,
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/useComboboxToggle.ts:51]: TS complained.

File: packages/components/src/Form/Inputs/Combobox/utils/useInputEvents.ts [hunk 14]
```diff
@@ -68,7 +82,7 @@ export function useInputEvents<
     data: { lastActionType },
     inputElement,
     openOnFocus,
-    persistSelectionRef,
+    persistSelectionPropRef,
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/useInputEvents.ts:85]: Name change for consistency.

File: packages/components/src/Form/Inputs/Combobox/utils/useOptionEvents.ts [hunk 15]
```diff
@@ -54,6 +56,10 @@ export function useOptionEvents<
       }
     }
     transition && transition(ComboboxActionType.SELECT_WITH_CLICK, { option })
+    if (closeOnSelectPropRef && closeOnSelectPropRef.current) {
+      // Closing an opened list
+      transition && transition(ComboboxActionType.ESCAPE)
+    }
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Combobox/utils/useOptionEvents.ts:62]: Adding this to explicitly close the list after changing the `stateChart` to not close it by default.

File: packages/components/src/Form/Inputs/InputChips/InputChipsBase.tsx [hunk 16]
```diff
@@ -66,7 +66,7 @@ export const InputChipsBaseInternal = forwardRef(
 
     function handleKeyDown(e: KeyboardEvent<HTMLInputElement>) {
       onKeyDown && onKeyDown(e)
-      if (e.key === 'Backspace') {
+      if (e.key === 'Backspace' && !e.defaultPrevented) {
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/InputChips/InputChipsBase.tsx:69]: For `SelectMulti`'s `removeOnBackspace={false}` behavior.

File: packages/components/src/Form/Inputs/Select/Select.tsx [hunk 17]
```diff
@@ -98,72 +81,6 @@ export interface SelectProps
...
-const SelectOptionGroup = ({ options, title }: SelectOptionGroupProps) => (
-  <Box py="xxsmall">
-    <SelectOptionGroupTitle>
-      <span />
-      {title}
-    </SelectOptionGroupTitle>
-    {options.map(renderOption)}
-  </Box>
-)
-
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Select/Select.tsx:166]: Moving to `SelectOptions.tsx`.

File: packages/components/src/Form/Inputs/Select/Select.tsx [hunk 18]
```diff
@@ -235,20 +153,7 @@ const SelectComponent = forwardRef(
...
-                  <SelectOptionGroup key={index} {...optionAsGroup} />
-                ) : (
-                  renderOption(option as SelectOptionObject, index)
-                )
-              })
-            ) : (
-              <ListItem fontSize="small" px="medium" py="xxsmall">
-                No options
-              </ListItem>
-            )}
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Select/Select.tsx:251]: Moving to `SelectOptions.tsx`.

File: packages/components/src/Form/Inputs/Select/Select.tsx [hunk 19]
```diff
@@ -98,72 +81,6 @@ export interface SelectProps
...
-  return value
-    ? { label: getComboboxText(value, flattenedOptions), value }
-    : undefined
-}
-
-function getFirstOption(options: SelectOptionProps[]): SelectOptionObject {
-  const optionAsGroup = options[0] as SelectOptionGroupProps
-  if (optionAsGroup.title) return optionAsGroup.options[0]
-  return options[0] as SelectOptionObject
-}
```
Comment (mdodgelooker) [inline packages/components/src/Form/Inputs/Select/Select.tsx:125]: Moving to `Select/utils/option.ts`.

File: packages/components/src/utils/usePopper.ts [hunk 20]
```diff
@@ -120,6 +120,9 @@ export function usePopper({
         mergedOptions
       )
     }
+    return () => {
+      popperInstanceRef.current && popperInstanceRef.current.destroy()
+    }
```
Comment (mdodgelooker) [inline packages/components/src/utils/usePopper.ts:125]: Fixes a React "setting state on an unmounted component" console error.

File: packages/components/src/Form/Inputs/Select/Select.test.tsx [hunk 21]
```diff
@@ -112,7 +165,102 @@ describe('<Select/>', () => {
...
-    expect(handleChange).toHaveBeenCalledWith('something')
+
+    const onChangeArg = name === 'SelectMulti' ? ['something'] : 'something'
+    expect(handleChange).toHaveBeenCalledWith(onChangeArg)
+    // Resolves "act" warning
+    await wait()
+  })
+
+  describe('no options', () => {
+    test.each([
```
Comment (dbchristopher) [inline packages/components/src/Form/Inputs/Select/Select.test.tsx:176]: I've never seen this test feature before! Nice way to ensure symmetry between Select and SelectMulti tests

**Comment by lukelooker** [comment]
Needs CHANGELOG update :)

**Comment by mdodgelooker** [comment]
@dbchristopher and @lukelooker you guys are letting me off easy 😉– I just updated the changelog and rebased.

**Pull request merged**